File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/D-Шарики/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    nums: iter[int] = map(int, sys.stdin.readline().rstrip().split())
    stack: list[list[int, int]] = []
    deletions: int = 0
    for _ in range(n):
        num: int = next(nums)
        if not stack or stack[-1][0] != num:
            if stack and stack[-1][-1] > 2:
                deletions += stack.pop()[-1]
                if stack and stack[-1][0] == num:
                    stack[-1][-1] += 1
                else: stack.append([num, 1])
            else: stack.append([num, 1])
        else: stack[-1][-1] += 1
    if stack and stack[-1][-1] > 2: deletions += stack[-1][-1]
    sys.stdout.write('{}\n'.format(deletions))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/D-Шарики/solution.cpp
#include <iostream>
#include <vector>

using namespace std;


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    scanf("%d", &n);
    vector<pair<int, int>> stack;
    int deletions = 0;
    for (int i = 0; i < n; ++i) {
        int num;
        cin >> num;
        if (stack.empty() || stack.back().first != num) {
            if (!stack.empty() && stack.back().second > 2) {
                deletions += stack.back().second;
                stack.pop_back();
                if (!stack.empty() && stack.back().first == num) {
                    ++stack.back().second;
                } else stack.push_back({num, 1});
            } else stack.push_back({num, 1});
        } else ++stack.back().second;
    }
    if (!stack.empty() && stack.back().second > 2) {
        deletions += stack.back().second;
    }
    printf("%d\n", deletions);
}

 
int main() {
    solution();
}